Python API improvements

mluessi opened the pull request:
Several improvements to the Python API:
- It is now possible to directly instantiate UAVOs in Python with the correct default values from the xml file. E.g.: `sparky = uavo.UAVO_HwSparky2()`
- ENUMs are included in the Python UAVOs (including correct parsing of shared definitions) and are e.g. used to produce more informative representations of the objects . E.g.:

```
    print sparky
    UAVO_HwSparky2(name=UAVO_HwSparky2, time=0, uavo_id=609948424, CoordID=0,
    RcvrPort=Disabled(0), MainPort=Disabled(0), FlexiPort=Disabled(0), USB_HIDPort=USBTelemetry(0), 
    USB_VCPPort=Disabled(4), DSMxMode=Autodetect(0), Radio=Disabled(0), MaxRfSpeed=64000(3), 
    MaxRfPower=125(1), MinChannel=0, MaxChannel=250, GyroRange=2000(3),  AccelRange=8G(2),
    MPU9250Rate=500(3), MPU9250GyroLPF=184(0), MPU9250AccelLPF=184(1), VTX_Ch=1(0),
    Magnetometer=Internal(0), ExtMagOrientation=Top0degCW(0))
```
- Python scripts using the githash to get the UAVO definitions can now be run from any directory (previously, one had to be in the TL root directory)
- Fixed a bug that sometimes caused the TCP telemetry to fail due to a "Resource temporarily unavailable" error (not sure why this error occurred, it shouldn't be possible due to the use of `select`).

mluessi committed:
Python API improvements

peabody124 commented:
Oh awesome. I'm super stoked about enums. That will make code so much clearer and it much easier to see things like flight mode

peabody124 commented:
Awesome. Im very stoked about enums

mlyle commented on python/taulabs/telemetry.py:
```diff
@@ -323,13 +323,19 @@ def _do_io(self, finish_time):
-            chunk = os.read(self.fd, 1024)
-            if chunk == '':
-                raise RuntimeError("stream closed")
+            try:
```
I'm a little concerned about this-- worries me it's masking a bug with when things are ready for reading.  EAGAIN after they told us it's ready to read?  How odd.  Maybe I am calling this at inappropriate times.

BrainFPV replied:
Yeah.. I have no idea how this can occur. It may be a bug in Python or something.

mluessi replied:
PS: I never saw it when parsing log files, only when connecting to the POSIX simulation using TCP

peabody124 replied:
Oh that's interesting. Was it when the simulator crashed?

mlyle replied:
I know we don't like TODO comments but I hate the idea of blocking this PR / a useful fix on this anomaly.  Maybe put TODO: in front of the above comment and call it good.

mluessi replied:
@peabody124 no, the simulator works normal and successive read calls are successful. I can add a note to the code.

mlyle commented on python/taulabs/uavo.py:
```diff
@@ -80,6 +85,29 @@ def from_bytes(cls, data, timestamp, instance_id, offset=0):
+        for field in self._fields:
+            field_value = getattr(self, field)
+            # Show ENUMs in a fancy way (string and value)
+            if hasattr(self, 'ENUM_' + field):
```
I think I'd prefer having an array of enums based on field index than all of this introspection and string manipulation.  But not a dealbreaker at all.

mluessi replied:
The ENUMs are also useful for setting values. E.g., one can use `UAVO_HwSparky2.ENUM_FlexiPort.Disabled` to get a specific value when manipulating UAVOs in code.

mlyle replied:
Sure, I'm not saying they should go away-- they should definitely be accessible symbolically.

mluessi replied:
Just to make sure, you are suggesting to add a list / tuple with the same length as `_fields` that also contains the enums (and `None` for things that are not enums)? Performance-wise, I'm not sure if this is needed, as the `__repr__` method is only used when printing objects and AFAIK `hasattr` is quite fast.

mlyle replied:
It's probably not needed.  I just don't really like introspection when it's not completely necessary.  Anyways, not a dealbreaker at all.

mlyle commented on python/taulabs/uavo.py:
```diff
@@ -183,18 +212,64 @@ def make_class(xml_file):
+                try:
+                    parent_class = globals()['UAVO_' + parent_name[0]]
+                    parent_options = getattr(parent_class, 'ENUM_' + field_name)
+                    if len(info['options']) == 0:
```
I'm really glad you handled this case.  The only additional thing needed is to adjust the hash to handle inheritance / sparse options.

mluessi replied:
So the hash in Python is currently broken? This was not caused by this PR, correct? What would need to be done to fix it?

mlyle replied:
The hash for whenever there's a sparse enum includes the next index before there's a gap.  So it only affects the hw* objects.

```
        if(info->fields[n]->type == FIELDTYPE_ENUM) {
            QStringList options = info->fields[n]->options;
            int nextIdx = 0;

            for (int m = 0; m < options.length(); m++) {
                int idx = findOptionIndex(info->fields[n], m);

                if (idx < 0) abort();

                // Not contiguous options.  Update with next value.
                if (idx != nextIdx) {
                    hash = updateHash((quint32) idx, hash);
                }

                nextIdx = idx+1;

                hash = updateHash(options[m], hash);
            }
```

is how it works on the C side.  now that you understand enums it's possible to fix.   It's issue #1804

mluessi replied:
Thanks, I will look into this and add it.

mlyle commented on python/taulabs/uavo.py:
```diff
@@ -269,10 +344,25 @@ class tmpClass(namedtuple(name, tuple_fields), UAVTupleClass):
+    # Set the default values for the constructor
+    defaults = [name, 0, uavo_id]
+    defaults.extend(f['defaultvalue'] for f in fields)
+    tuple_class.__new__.__defaults__ = tuple(defaults)
```
Super slick.  I've wanted to do this.  Thank you.

mluessi replied:
Some serious Python magic :)

mluessi committed:
Python API: fix uavo hash

mluessi commented:
Fixed the hash calculation. I had introduced some more bugs by "forgetting" that dicts are not ordered and by using a regex to modify the enum options (so they could be used as attributes in code). The enums and the reverse mappings are now just dicts and can be accessed like this:

    print uavo.UAVO_HwSparky2.ENUM_Magnetometer['Internal']
    0
    print uavo.UAVO_HwSparky2.ENUMR_Magnetometer[2]
    ExternalAuxI2C

Fixes #1804

mluessi commented on python/taulabs/uavo.py:
```diff
@@ -269,10 +346,25 @@ class tmpClass(namedtuple(name, tuple_fields), UAVTupleClass):
+            mapping = dict((key, value) for key, value in enum.iteritems())
+            reverse_mapping = dict((value, key) for key, value in enum.iteritems())
+            setattr(tuple_class, 'ENUM_' + field['name'], mapping)
+            setattr(tuple_class, 'ENUMR_' + field['name'], reverse_mapping)
```
Instead of using a two dicts, we could just use one dict that contains the mapping from integer to string and vice versa. May be easier to use. I.e., `ENUM_FlightMode[0]` would return a string while `ENUM_FlightMode['Manual']` would return an int. Thoughts?

mlyle replied:
I think whaty ou've got is great.

mluessi committed:
Python: add comment about read failure

mluessi commented:
Added a comment about the read failure.

mluessi commented:
Is there anything else that needs to be done? If not, this should be good to merge.

kubark42 commented:
@mluessi The branch has merge conflicts.

mluessi commented:
@kubark42 rebased

kubark42 commented on python/taulabs/telemetry.py:
```diff
@@ -327,13 +327,20 @@ def _do_io(self, finish_time):
+                    raise RuntimeError("stream closed")
 
-            self.recv_buf = self.recv_buf + chunk
+                self.recv_buf = self.recv_buf + chunk
```
Ugh, python. This is exactly why I hate semantic white-spacing. :p

mluessi replied:
? this is correct (needs be indented)

kubark42 replied:
Yah, realized it's correct. But it's annoyingly hard to disambiguate between accidental IDE style changes and intentional scope changes. No matter, just ranting.

kubark42 commented:
LGTM, and everyone seems to like it. Win!
